pkg/controller/deployment/deployment_controller.go

```diff
@@ -591,14 +580,91 @@ func (dc *DeploymentController) syncRollingUpdateDeployment(deployment extension
+	if err != nil {
+		return err
+	}
+	return dc.proportionalScaling(deployment, newRS, oldRSs)
+}
+
+// proportionalScaling should run only on scaling events or when a deployment is paused
+// and not during the normal rollout process.
+func (dc *DeploymentController) proportionalScaling(deployment extensions.Deployment, newRS *extensions.ReplicaSet, oldRSs []*extensions.ReplicaSet) error {
+	allRSs := append(controller.FilterActiveReplicaSets(oldRSs), newRS)
+
+	// If we scale a paused or a recreate deployment, it may not have a new replica set...
+	if newRS != nil {
+		// No old replica sets with pods; scale new replica set to the full count.
+		if len(controller.FilterActiveReplicaSets(oldRSs)) == 0 {
+			glog.V(4).Infof("Scaling new replica set %q from %d to %d", newRS.Name, newRS.Spec.Replicas, deployment.Spec.Replicas)
+			if _, _, err := dc.scaleReplicaSetAndRecordEvent(newRS, deployment.Spec.Replicas, deployment); err != nil {
+				return err
+			}
+			return dc.syncDeploymentStatus(allRSs, newRS, deployment)
+		}
+	}
+
+	// There are old replica sets with pods. We need to proportionally scale all rs
+	// (new and old).
+	switch deployment.Spec.Strategy.Type {
+	case extensions.RecreateDeploymentStrategyType:
+		// Proportional scaling is not valid for recreate deployments. There should be at
+		// most one replica set running at any given point. If we don't scale up the new
+		// replica set above, then we should ignore scaling events during the rollout process.
+	case extensions.RollingUpdateDeploymentStrategyType:
+		// Scaling proportionally is valid only for rolling deployments.
+		sort.Sort(controller.BySize(allRSs))
+
+		totalReplicas := deploymentutil.GetReplicaCountForReplicaSets(allRSs)
+		totalAdded := 0
+		// Number of additional pods that can be either added or removed from the total pod count.
+		additional := deployment.Spec.Replicas - totalReplicas
+		if additional == 0 {
+			return dc.syncDeploymentStatus(allRSs, newRS, deployment)
+		}
+		// Scale proportionally.
+		for i := range allRSs {
+			rs := allRSs[i]
+			if rs == nil {
+				continue
+			}
+			proportion := 0
+			if additional > 0 {
+				// Scaling up...
+				proportion = integer.IntMin(integer.RoundToInt((float64(rs.Spec.Replicas)/float64(totalReplicas))*float64(additional)), additional-totalAdded)
+			} else {
+				// Scaling down...
+				proportion = integer.IntMax(integer.RoundToInt((float64(rs.Spec.Replicas)/float64(totalReplicas))*float64(additional)), additional-totalAdded)
+			}
+			// This can be either positive (scale up) or negative (scale down).
+			if proportion == 0 {
+				continue
+			}
+			glog.V(4).Infof("Proportionally scaling replica set %q from %d to %d", rs.Name, rs.Spec.Replicas, rs.Spec.Replicas+proportion)
+			if _, _, err := dc.scaleReplicaSetAndRecordEvent(rs, rs.Spec.Replicas+proportion, deployment); err != nil {
+				glog.Infof("Cannot scale replica set %q to %d replicas: %v", rs.Name, rs.Spec.Replicas+proportion, err)
```

Comment:
Replica sets that were updated will be reported in the status because we don't error out. We probably falsely would report that the latest was updated... Maybe we should error out. A scaling event that errored would be retried in the next sync loop because deployment spec should eventually persist.